src/components/elements/auto-carousel.js

```diff
@@ -1,92 +1,112 @@
+        transform: translateX(0);
+    }
+    100% {
+        transform: translateX(-${count_child}%);
     }
 `
-const CarouselWrapper = styled.div`
-    max-width: 9.6rem;
-    margin-right: 3.2rem;
+const ItemContainer = styled.div`
+    padding: 0 ${props => props.padding / 2}px;
+    animation-name: ${props =>
+        props.shoudl_carousel_move ? move_items(props.count_child * 100) : ''};
+    animation-duration: ${props => props.transition_duration}ms;
+    animation-fill-mode: both;
+    animation-iteration-count: infinite;
+    animation-timing-function: linear;
 `
-const Icons = ({ icon_array }) =>
-    icon_array.map(Element => (
-        <CarouselWrapper key={Element.key}>
-            <Element.Component />
-        </CarouselWrapper>
-    ))
-
-class AutoCarousel extends React.Component {
-    interval_ref = undefined
-    timeout_ref = undefined
+class AutoCarousel extends React.PureComponent {
     my_ref = React.createRef()
-    state = {
-        position: 0,
-        transition: true,
+    constructor(props) {
+        super(props)
+        this.state = {
+            transition: false,
+            items: [],
+            carousel_width: 0,
+            shoudl_carousel_move: false,
+        }
     }
+    // every time you observe this the carousel it restart from first component
     handler = entries => {
         let entry
         for (entry of entries) {
             if (entry.isIntersecting) {
-                this.interval_ref = window.setInterval(
-                    this.handleInterval,
-                    this.props.transition_delay,
-                )
+                this.setState({
+                    shoudl_carousel_move: true,
+                })
             } else {
-                window.clearTimeout(this.timeout_ref)
-                window.clearInterval(this.interval_ref)
+                this.setState({
+                    shoudl_carousel_move: false,
+                })
             }
         }
     }
-
-    handleInterval = () => {
-        this.setState({
-            transition: true,
-            position: this.state.position - this.props.icon_width,
-        })
-        this.timeout_ref = setTimeout(() => {
-            const new_array = this.props.components
-            const new_array_icon = new_array.shift()
-            new_array.push(new_array_icon)
-            this.setState(
-                {
-                    transition: false,
-                    position: 0,
-                },
-                this.props.onChange(new_array),
-            )
-        }, this.props.transition_delay - 100)
-    }
     componentDidMount() {
-        const node = this.my_ref.current
         let observer = new IntersectionObserver(this.handler)
-        observer.observe(node)
+        observer.observe(this.my_ref.current)
+    }
+    static getDerivedStateFromProps(props, state) {
+        if (state.items.length === 0) {
+            const newItems = [...props.children, ...props.children].map(
```

Comment:
shouldn't it be just single `...props.children` ?

Reply:
No, If we use single `...props.children` when it reaches the end of the `auto-carousel` we will see empty space. so we should render 2 of them right after each other.